params/types/coregeth/chain_config_configurator.go

```diff
@@ -516,51 +516,147 @@ func (c *CoreGethChainConfig) SetEIP4399Transition(n *uint64) error {
 
 // EIP3651: Warm COINBASE
 func (c *CoreGethChainConfig) GetEIP3651TransitionTime() *uint64 {
+	if c.EIP3651FBlock != nil && c.EIP3651FBlock.Cmp(big.NewInt(0)) == 0 {
+		return newU64(0)
+	}
 	return c.EIP3651FTime
 }
 
 func (c *CoreGethChainConfig) SetEIP3651TransitionTime(n *uint64) error {
 	c.EIP3651FTime = n
+	if n != nil && *n == 0 {
+		c.EIP3651FBlock = big.NewInt(0)
+	}
 	return nil
 }
 
 // GetEIP3855TransitionTime EIP3855: PUSH0 instruction
 func (c *CoreGethChainConfig) GetEIP3855TransitionTime() *uint64 {
+	if c.EIP3855FBlock != nil && c.EIP3855FBlock.Cmp(big.NewInt(0)) == 0 {
+		return newU64(0)
+	}
 	return c.EIP3855FTime
 }
 
 func (c *CoreGethChainConfig) SetEIP3855TransitionTime(n *uint64) error {
 	c.EIP3855FTime = n
+	if n != nil && *n == 0 {
+		c.EIP3855FBlock = big.NewInt(0)
+	}
 	return nil
 }
 
 // GetEIP3860TransitionTime EIP3860: Limit and meter initcode
 func (c *CoreGethChainConfig) GetEIP3860TransitionTime() *uint64 {
+	if c.EIP3860FBlock != nil && c.EIP3860FBlock.Cmp(big.NewInt(0)) == 0 {
+		return newU64(0)
+	}
 	return c.EIP3860FTime
 }
 
 func (c *CoreGethChainConfig) SetEIP3860TransitionTime(n *uint64) error {
 	c.EIP3860FTime = n
+	if n != nil && *n == 0 {
+		c.EIP3860FBlock = big.NewInt(0)
+	}
 	return nil
 }
 
 // GetEIP4895TransitionTime EIP4895: Beacon chain push withdrawals as operations
 func (c *CoreGethChainConfig) GetEIP4895TransitionTime() *uint64 {
+	if c.EIP4895FBlock != nil && c.EIP4895FBlock.Cmp(big.NewInt(0)) == 0 {
+		return newU64(0)
+	}
 	return c.EIP4895FTime
 }
 
 func (c *CoreGethChainConfig) SetEIP4895TransitionTime(n *uint64) error {
 	c.EIP4895FTime = n
+	if n != nil && *n == 0 {
+		c.EIP4895FBlock = big.NewInt(0)
+	}
 	return nil
 }
 
 // GetEIP6049TransitionTime EIP6049: Deprecate SELFDESTRUCT
 func (c *CoreGethChainConfig) GetEIP6049TransitionTime() *uint64 {
+	if c.EIP6049FBlock != nil && c.EIP6049FBlock.Cmp(big.NewInt(0)) == 0 {
+		return newU64(0)
+	}
 	return c.EIP6049FTime
 }
 
 func (c *CoreGethChainConfig) SetEIP6049TransitionTime(n *uint64) error {
 	c.EIP6049FTime = n
+	if n != nil && *n == 0 {
+		c.EIP6049FBlock = big.NewInt(0)
+	}
+	return nil
+}
+
+// Shanghai by block
+// EIP3651: Warm COINBASE
+func (c *CoreGethChainConfig) GetEIP3651Transition() *uint64 {
```

Comment:
I like how the setters and getters of EIPs that can be activated with both time and block number check to each other, although I am a bit worried to what happens if both are being set by mistake and if the user will notice it.

Do you think that it is better if we take precedence of one over the other? For example, first check time then block number no matter which getter is being used.

Currently the siblings is only checked if enabled on genesis block, though some developers might mistakenly pass 0 values to some EIPs when they init chains using custom genesis.

Reply:
> Do you think that it is better if we take precedence of one over the other? For example, first check time then block number no matter which getter is being used.

The problem is that each getter can only represent the time OR block value; a single getter can't reconcile between the two kinds of activation values (except for the special `0` case, where `time:0 -> block:0`; even `block:0 -> time:0` is not logically sound, since genesis blocks can, and do, have non-`0` timestamps).

And I think the problem persists, were precedence to be established, with developers mistakenly passing `nil` values (true zero-values), the same with `0` values. 

But I'm not sure I've perfectly understood your proposal, nor that I've even provided a watertight reason -- maybe you can show a pseudocode or snippet of what you're considering?

Reply:
@meowsbits I removed this logic as it's not needed. There was a copy/paste typo which caused all this. Here is the fix [26265c7](https://github.com/etclabscore/core-geth/pull/561/commits/26265c73c011f2e678f6aff4fc99db114cc5d4bf).